src/Domain/CreateInitialElement.cpp

```diff
@@ -3,52 +3,124 @@
+    const auto& refinement_of_neighbor =
+        initial_refinement[block_neighbor.id()];
+
+    // Check whether each dimension will have multiple neighbors
+    // because the neighboring block is more refined.  For dimensions
+    // that have only one neighbor, adjust the SegmentId to that of
+    // the neighbor if necessary.
+    auto neighbor_is_refined = make_array<VolumeDim>(false);
+    for (size_t d = 0; d < VolumeDim; ++d) {
+      switch (static_cast<int>(gsl::at(refinement_of_neighbor, d)) -
+              static_cast<int>(gsl::at(segment_ids_of_unrefined_neighbor, d)
+                                   .refinement_level())) {
+        case 0:
+          break;
+        case 1:
+          gsl::at(neighbor_is_refined, d) = true;
+          break;
+        case -1:
+          gsl::at(segment_ids_of_unrefined_neighbor, d) =
+              gsl::at(segment_ids_of_unrefined_neighbor, d).id_of_parent();
+          break;
+        default:
+          ERROR("Refinement levels "
+                << gsl::at(refinement_of_neighbor, d) << " and "
+                << gsl::at(segment_ids_of_unrefined_neighbor, d)
+                       .refinement_level()
+                << " in blocks " << block_neighbor.id() << " and " << block.id()
+                << " differ by more than one.");
+      }
+    }
+
+    // Refinement in the dimension perpendicular to the interface
+    // never causes additional neighbors, even if the neighboring
+    // elements are smaller.
+    {
+      auto& perpendicular_segment = gsl::at(segment_ids_of_unrefined_neighbor,
+                                            direction_in_neighbor.dimension());
+      auto& perpendicular_refinement =
+          gsl::at(neighbor_is_refined, direction_in_neighbor.dimension());
+      if (perpendicular_refinement) {
+        perpendicular_segment =
+            perpendicular_segment.id_of_child(direction_in_neighbor.side());
+        perpendicular_refinement = false;
+      }
+      perpendicular_segment = perpendicular_segment.id_if_flipped();
+    }
+
+    // Consider all possible neighbors by looping over all elements of
+    // the product set {Lower, Upper}^VolumeDim, checking whether each
+    // actually describes a neighbor.  For dimensions where the
+    // interface is not split, we arbitrarily call all neighbors Lower
+    // and consider any Upper neighbors invalid.
+    std::unordered_set<ElementId<VolumeDim>> neighbor_ids;
+    for (IndexIterator<VolumeDim> index(Index<VolumeDim>(2)); index; ++index) {
+      std::array<SegmentId, VolumeDim> segment_ids_of_neighbor;
+      for (size_t d = 0; d < VolumeDim; ++d) {
+        const auto& unrefined_segment =
+            gsl::at(segment_ids_of_unrefined_neighbor, d);
+        auto& segment = gsl::at(segment_ids_of_neighbor, d);
+        if (not gsl::at(neighbor_is_refined, d)) {
+          if ((*index)[d] == 0) {
+            segment = unrefined_segment;
+          } else {
+            goto next_index;
```

Comment:
there's no way to review this PR without commenting on this line :) is there no way to `continue` an outer loop in c++?

Reply:
Nope.  C(++) has the worst loop control of any major language I know.  Including bash.

Reply:
My understanding is that the only way to evade the `goto` would be to add extra `bool` flags and make the loop conditions more complicated. I think it is generally possible to avoid `goto`s, but is certainly more labor-intensive and intricate -- I suppose it depends entirely on how much we want to subscribe to the 'no gotos' religion of Dijkstra.
I'd be in favor of permitting limited, local, strictly forward gotos, which is what is used here.